Move ClipExtension to @deck.gl/extensions

zakjan opened the pull request:
Fixes #5725

Pessimistress commented:
Thanks!

TODO:

- [ ] add ClipExtension documentation
- [ ] update MVTLayer documentation to mention the new dependency
- [ ] add to upgrade guide that geo-layers now has new dependency on extensions

ibgreen commented:
@zakjan This looks good - are you able to address the remaining items @Pessimistress listed or do you need some pointers?

zakjan commented:
Docs added for ClipExtension and clipBounds, added as dependency for MVTLayer, added to upgrade guide. Let me know if it fits with common docs writing style.

I would need help with documenting clipByInstance though. It's not used anywhere, I'm not sure what's the usecase.

Also, I found that similar implicit clipping is used in HeatmapLayer. Is it duplicate, or a different usecase? https://github.com/visgl/deck.gl/blob/26bcb63f238158ff2bdbf629787092bd92d6c88d/modules/aggregation-layers/src/heatmap-layer/heatmap-layer.js#L401-L407

ibgreen commented:
> Also, I found that similar implicit clipping is used in HeatmapLayer. Is it duplicate, or a different usecase?

Interesting observation. I am not current on the heatmap layer code, but generally I have a sense that it has been a rather big separate piece of code that is heavier to maintain than I would prefer, and it would merit from being more "componentized" and aligned with other parts of the code. Using the ClipExtension there could be a good inroad, if it makes sense.

I can imagine that some of the weather focused layer work you are setting out to do would have parallels with the heatmap layer (rendering into textures etc) and there could be multiple refactor / code sharing opportunities for someone who was able to get familiar with the heatmap layer code base.

Definitely a separate PR. @Pessimistress likely has more insights as to what would be feasible.

Pessimistress commented:
HeatmapLayer does not perform clipping in the shader. The texture is dynamically generated to fill the viewport. The referenced code is totally unrelated to the use case of the ClipExtension.

Pessimistress commented on docs/api-reference/extensions/clip-extension.md:
```diff
@@ -0,0 +1,59 @@
+import {BitmapLayer} from '@deck.gl/layers';
+import {ClipExtension} from '@deck.gl/extensions';
+
+const layer = new BitmapLayer({
```
I recommend using a GeoJsonLayer as example. The code here does not illustrate why clipping is useful.

Pessimistress commented on docs/api-reference/extensions/clip-extension.md:
````diff
@@ -0,0 +1,59 @@
+```bash
+npm install deck.gl
+# or
+npm install @deck.gl/core @deck.gl/layers @deck.gl/extensions
````
layers is not a dependency of extensions

Pessimistress commented on docs/api-reference/extensions/clip-extension.md:
```diff
@@ -0,0 +1,59 @@
+<script src="https://unpkg.com/deck.gl@^7.0.0/dist.min.js"></script>
+<!-- or -->
+<script src="https://unpkg.com/@deck.gl/core@^7.0.0/dist.min.js"></script>
+<script src="https://unpkg.com/@deck.gl/layers@^7.0.0/dist.min.js"></script>
```
Ditto

Pessimistress commented on docs/api-reference/extensions/clip-extension.md:
```diff
@@ -0,0 +1,59 @@
+
+## Layer Properties
+
+When added to a layer via the `extensions` prop, the `ClipExtension` requires the `clipBounds` prop of the layer to contain the rectangular bounds to be used for clipping the rendered region.
```
Each prop should have its own header, follow the convention in https://github.com/visgl/deck.gl/blob/master/docs/api-reference/extensions/data-filter-extension.md#layer-properties

Pessimistress reviewed (commented):
`clipByInstance` controls whether an object is clipped by its anchor (e.g. icon, point) or by its geometry (e.g. path, polygon). If not specified, it is deduced from whether there is an attribute called `instancePositions`. The user might want to override this behavior if:
- the anchor attribute goes by some other name
- to clip an anchored layer by geometry, like the text layer

zakjan commented:
Thanks @Pessimistress, initially I used Fp64Extension as a template. Docs updated.
